noir_stdlib/src/lib.nr

```diff
@@ -105,17 +105,48 @@ where
     AsPrimitive::as_(x.as_() + 340282366920938463463374607431768211456 - y.as_())
 }
 
-pub fn wrapping_mul<T>(x: T, y: T) -> T
+fn wrapping_mul<T>(x: T, y: T) -> T
 where
-    T: AsPrimitive<Field>,
-    Field: AsPrimitive<T>,
+    T: ops::arith::WappingMul,
 {
-    AsPrimitive::as_(x.as_() * y.as_())
+    x.wrapping_mul(y)
 }
 
 #[builtin(as_witness)]
 pub fn as_witness(x: Field) {}
 
+global two_pow_64: Field = 0x10000000000000000;
+/// Splits a 128 bits number into two 64 bits limbs
+unconstrained fn split64(x: Field) -> (u64, u64) {
+    let lo = x as u64;
+    let hi = ((x - lo as Field) / two_pow_64) as u64;
+    (lo, hi)
+}
+
+/// Split a 128 bits number into two 64 bits limbs
+/// It will fail if the number is more than 128 bits
+fn split_into_64_bit_limbs(x: Field) -> (u64, u64) {
+    // Safety: the limbs are constrained below
+    let (x_lo, x_hi) = unsafe { split64(x) };
+    assert(x as Field == x_lo as Field + x_hi as Field * two_pow_64);
+    (x_lo, x_hi)
+}
+
+#[field(bn254)]
+pub fn wrapping_mul128(x: Field, y: Field) -> Field {
+    let (x_lo, x_hi) = split_into_64_bit_limbs(x);
+    let (y_lo, y_hi) = split_into_64_bit_limbs(y);
+    // Multiplication using the limbs:(x_lo + 2**64*x_hi)*(y_lo + 2**64*y_hi)=x_lo*y_lo+...
+    // and skipping the terms over 2**128
+    // Working with u64 limbs ensures that we cannot overflow the field modulus.
+    let low = x_lo as Field * y_lo as Field;
+    let lo = low as u64 as Field;
+    let carry = (low - lo) / two_pow_64;
+    let high = x_lo as Field * y_hi as Field + x_hi as Field * y_lo as Field + carry;
+    let hi = high as u64 as Field;
+    (lo + two_pow_64 * hi)
+}
```

Comment:
These shouldn't be at the crate root either.

Reply:
Done